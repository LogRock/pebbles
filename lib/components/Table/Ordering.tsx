import React, { FC, useContext } from "react";
import { ThemeContext } from "styled-components";
import { OrderingEnum, OrderingProps } from "./Table.types";

const Ordering: FC<OrderingProps> = ({ currentOrder }) => {
  const theme = useContext(ThemeContext);

  if (currentOrder === OrderingEnum.ASC)
    return (
      <svg
        width="12"
        height="7"
        viewBox="0 0 12 7"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
        style={{ marginTop: 6 }}
      >
        <path
          d="M11 6L6 1L1 6"
          stroke="#6B7280"
          strokeWidth="1.67"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
    );

  if (currentOrder === OrderingEnum.DESC)
    return (
      <svg
        width="12"
        height="7"
        viewBox="0 0 12 7"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
        style={{ marginTop: 6 }}
      >
        <path
          d="M11 1L6 6L1 1"
          stroke="#6B7280"
          strokeWidth="1.67"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
    );

  return (
    <svg
      width="11"
      height="16"
      viewBox="0 0 11 16"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      style={{ verticalAlign: "middle" }}
    >
      <>
        <path
          d="M1.375 5L5.375 1L9.375 5M9.375"
          stroke={theme.colors.neutral[400]}
          strokeWidth="1.67"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
        <path
          d="M1.375 5M9.375 11L5.375 15L1.375 11"
          stroke={theme.colors.neutral[400]}
          strokeWidth="1.67"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </>
    </svg>
  );
};

export default Ordering;
