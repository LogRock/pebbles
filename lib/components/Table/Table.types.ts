import React from "react";
import { BaseItemType } from "../Select/Select.types";

export interface TableHeaderProps extends React.ComponentPropsWithoutRef<"th"> {
  showOrdering?: boolean;
  currentOrder?: OrderingEnum;
  center?: boolean;
}

export enum OrderingEnum {
  // eslint-disable-next-line no-unused-vars
  ASC = "ASC",
  // eslint-disable-next-line no-unused-vars
  DESC = "DESC",
}

export interface OrderingProps {
  currentOrder?: OrderingEnum;
}

export interface TableRowProps extends React.ComponentPropsWithoutRef<"tr"> {
  isNew?: boolean;
  isSelected?: boolean;
  addEmptyTDAtEnd?: boolean;
}

export interface TableFooterProps {
  showItemsCount?: boolean;
  itemsFrom?: string;
  itemsTo?: string;
  itemsTotal?: string;
  itemsFromToTotalFormatter?: (
    from: string,
    to: string,
    total: string
  ) => string;
  showGoToFirstPage?: boolean;
  goToFirstPageContent?: React.ReactNode;
  onGoToFirstPageClicked?: () => void;
  showGoToPreviousPage?: boolean;
  goToPreviousPageContent?: React.ReactNode;
  onGoToPreviousPageClicked?: () => void;
  showGoToNextPage?: boolean;
  goToNextPageContent?: React.ReactNode;
  onGoToNextPageClicked?: () => void;
  showGoToLastPage?: boolean;
  goToLastPageContent?: React.ReactNode;
  onGoToLastPageClicked?: () => void;
  itemsPerPage?: number;
  itemsPerPageOptions?: BaseItemType[];
  onSetItemsPerPage?: (item: BaseItemType) => void;
}
