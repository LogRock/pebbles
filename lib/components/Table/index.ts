export {
  default,
  TableBody,
  TableDatum,
  TableFooter,
  TableHead,
  TableHeader,
  TableRow,
} from "./Table";

export { OrderingEnum } from "./Table.types";

export type {
  OrderingProps,
  TableFooterProps,
  TableHeaderProps,
  TableRowProps,
} from "./Table.types";
