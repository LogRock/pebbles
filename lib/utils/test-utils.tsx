/* eslint-disable import/export */
import React from "react";
import { cleanup, render } from "@testing-library/react";
import { afterEach } from "vitest";
import { ThemeProvider } from "styled-components";
import { light as theme } from "../theme";

afterEach(() => {
  cleanup();
});

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => (
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    ),
    ...options,
  });

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
// override render export
export { customRender as render };
