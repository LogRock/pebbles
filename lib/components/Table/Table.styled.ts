import styled, { css } from "styled-components";
import Select from "../Select";
import { SelectItems } from "../Select/Select.styled";
import {
  overlineXSmallCSS,
  paragraphMediumCSS,
  paragraphSmallCSS,
  ParagraphXSmall,
  paragraphXSmallCSS,
} from "../Typography";

import { TableHeaderProps, TableRowProps } from "./Table.types";

export const Table = styled.table`
  border-spacing: 0;

  border-radius: 10px;

  box-shadow: ${({ theme }) => theme.shadows.medium};

  tbody tr::after,
  tfoot tr::after {
    content: "";

    position: absolute;
    top: 0;
    right: ${({ theme }) => theme.spacings.md};
    left: ${({ theme }) => theme.spacings.md};
    height: 1px;

    background-color: ${({ theme }) => theme.colors.neutral[200]};
  }

  tbody tr:hover {
    background-color: ${({ theme }) => theme.colors.neutral[100]} !important;
  }
`;

export const TableHead = styled.thead`
  /* stylelint-disable-next-line */
`;

export const TableBody = styled.tbody`
  /* stylelint-disable-next-line */
`;

export const TFoot = styled.tfoot`
  position: relative;
  ${paragraphXSmallCSS}

  width: 100%;

  background-color: ${({ theme }) => theme.colors.shades[0]};

  tr {
    td {
      padding-top: ${({ theme }) => theme.spacings.md};
      padding-right: ${({ theme }) => theme.spacings.big};
      padding-bottom: ${({ theme }) => theme.spacings.md};
      padding-left: ${({ theme }) => theme.spacings.big};

      text-align: right;
      vertical-align: middle;
    }
  }
`;

export const TH = styled.th<TableHeaderProps>`
  ${overlineXSmallCSS}

  padding: ${({ theme }) => theme.spacings.md};

  text-align: ${({ center }) => (center ? "center" : "left")};

  svg {
    height: 100%;
    margin-left: 10px;
  }
`;

export const TableDatum = styled.td<{ center?: boolean }>`
  ${paragraphMediumCSS}
  padding: ${({ theme }) => theme.spacings.md};

  text-align: ${({ center }) => (center ? "center" : undefined)};
`;

export const TR = styled.tr<TableRowProps>`
  position: relative;

  background-color: ${({ theme, isSelected }) =>
    isSelected ? theme.colors.primary[50] : theme.colors.shades[0]} !important;

  &:nth-child(even) {
    background-color: ${({ theme, isSelected }) =>
      !isSelected && theme.colors.neutral[50]} !important;
  }

  ${TableDatum} {
    font-weight: ${({ isNew }) => (isNew ? 600 : 400)} !important;
  }

  ${TableDatum}:first-of-type {
    ${({ isNew }) =>
      isNew
        ? css`
            &::before {
              content: " ";

              position: absolute;
              top: 0;
              bottom: 0;
              left: 0;
              width: 3px;

              border-top-right-radius: 3px;
              border-bottom-right-radius: 3px;

              background: ${({ theme }) => theme.colors.success[500]};
            }
          `
        : undefined}
  }
`;

export const New = styled.span`
  ${paragraphSmallCSS}
  padding: 4px 8px;

  border-radius: 100px;

  background-color: ${({ theme }) => theme.colors.success[50]};

  font-weight: 500;
`;

export const TablePaginationButton = styled.button`
  padding: 0 18px;

  border: none;

  background: none;

  vertical-align: middle;

  &:active {
    background-color: ${({ theme }) => theme.colors.neutral[100]};
  }
`;

export const TablePerPage = styled.div`
  display: flex;
  align-items: center;
  width: max-content;
`;

export const FooterLeftArea = styled.td`
  display: flex;
  align-items: center;
`;

export const TablePerPageSelector = styled(Select)`
  max-width: 120px;
  margin-right: ${({ theme }) => theme.spacings.md};

  input,
  button {
    height: 36px !important;

    font-size: 14px !important;
  }

  button {
    display: flex;
    flex-direction: column;
    justify-content: center;
  }

  ${SelectItems} {
    top: 48px;
  }
`;

export const TablePerPageLabel = styled(ParagraphXSmall)`
  margin-right: 8px;

  color: ${({ theme }) => theme.colors.neutral[500]};
  text-transform: uppercase;
`;
