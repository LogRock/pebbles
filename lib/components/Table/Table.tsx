import {
  mdiChevronDoubleLeft,
  mdiChevronDoubleRight,
  mdiChevronLeft,
  mdiChevronRight,
} from "@mdi/js";
import Icon from "@mdi/react";
import { isFunction } from "lodash";
import React, { FC, useContext, useEffect, useMemo, useState } from "react";
import { ThemeContext } from "styled-components";
import { BaseItemType } from "../Select/Select.types";
import Ordering from "./Ordering";
import {
  New,
  TableDatum,
  TablePaginationButton,
  TFoot,
  TH,
  TR,
  TablePerPage,
  TablePerPageLabel,
  TablePerPageSelector,
  ItemsCount,
} from "./Table.styled";
import {
  TableHeaderProps,
  TableRowProps,
  TableFooterProps,
} from "./Table.types";

export {
  Table as default,
  TableHead,
  TableBody,
  TableDatum,
  TFoot,
} from "./Table.styled";

export const TableRow: FC<TableRowProps> = ({
  isNew,
  children,
  addEmptyTDAtEnd,
  ...props
}) => {
  return (
    <TR {...props} isNew={isNew}>
      {children}
      {isNew ? (
        <TableDatum>
          <New>New</New>
        </TableDatum>
      ) : null}
      {addEmptyTDAtEnd ? <TableDatum></TableDatum> : null}
    </TR>
  );
};

export const TableHeader: FC<TableHeaderProps> = ({
  showOrdering,
  children,
  currentOrder,
  ...props
}) => {
  return (
    <TH {...props}>
      {children}
      {showOrdering && <Ordering currentOrder={currentOrder} />}
    </TH>
  );
};

export const TableFooter: FC<TableFooterProps> = ({
  showItemsCount,
  itemsFrom,
  itemsTo,
  itemsTotal,
  itemsFromToTotalFormatter,
  showGoToFirstPage,
  goToFirstPageContent,
  onGoToFirstPageClicked,
  showGoToPreviousPage,
  goToPreviousPageContent,
  onGoToPreviousPageClicked,
  showGoToNextPage,
  goToNextPageContent,
  onGoToNextPageClicked,
  showGoToLastPage,
  goToLastPageContent,
  onGoToLastPageClicked,
  itemsPerPageOptions,
  onSetItemsPerPage,
  children,
}) => {
  const theme = useContext(ThemeContext);

  const [pageLimit, setPageLimit] = useState(10);

  useEffect(() => {
    // console.log("pageLimit: ", pageLimit);
  }, [pageLimit]);

  const format =
    itemsFromToTotalFormatter && isFunction(itemsFromToTotalFormatter)
      ? itemsFromToTotalFormatter
      : (from: string, to: string, total: string) =>
          `${from}-${to} of ${total}`;

  const itemsCount = showItemsCount ? (
    <ItemsCount>
      {format(itemsFrom || "", itemsTo || "", itemsTotal || "")}
    </ItemsCount>
  ) : null;

  const goToFirstPage = useMemo(() => {
    if (showGoToFirstPage) {
      return (
        <TablePaginationButton
          onClick={
            isFunction(onGoToFirstPageClicked)
              ? onGoToFirstPageClicked
              : () => null
          }
        >
          {goToFirstPageContent || (
            <Icon
              path={mdiChevronDoubleLeft}
              size="18px"
              color={theme.colors.neutral[700]}
            />
          )}
        </TablePaginationButton>
      );
    }

    return "";
  }, [showGoToFirstPage, onGoToFirstPageClicked, goToFirstPageContent]);

  const goToPreviousPage = useMemo(() => {
    if (showGoToPreviousPage) {
      return (
        <TablePaginationButton
          onClick={
            isFunction(onGoToPreviousPageClicked)
              ? onGoToPreviousPageClicked
              : () => null
          }
        >
          {goToPreviousPageContent || (
            <Icon
              path={mdiChevronLeft}
              size="18px"
              color={theme.colors.neutral[700]}
            />
          )}
        </TablePaginationButton>
      );
    }

    return "";
  }, [
    showGoToPreviousPage,
    onGoToPreviousPageClicked,
    goToPreviousPageContent,
  ]);

  const goToNextPage = useMemo(() => {
    if (showGoToNextPage) {
      return (
        <TablePaginationButton
          onClick={
            isFunction(onGoToNextPageClicked)
              ? onGoToNextPageClicked
              : () => null
          }
        >
          {goToNextPageContent || (
            <Icon
              path={mdiChevronRight}
              size="18px"
              color={theme.colors.neutral[700]}
            />
          )}
        </TablePaginationButton>
      );
    }

    return "";
  }, [showGoToNextPage, onGoToNextPageClicked, goToNextPageContent]);

  const goToLastPage = useMemo(() => {
    if (showGoToLastPage) {
      return (
        <TablePaginationButton
          onClick={
            isFunction(onGoToLastPageClicked)
              ? onGoToLastPageClicked
              : () => null
          }
        >
          {goToLastPageContent || (
            <Icon
              path={mdiChevronDoubleRight}
              size="18px"
              color={theme.colors.neutral[700]}
            />
          )}
        </TablePaginationButton>
      );
    }

    return "";
  }, [showGoToLastPage, onGoToLastPageClicked, goToLastPageContent]);

  const itemsPerPageSelector = useMemo(() => {
    if (itemsPerPageOptions) {
      return (
        <TablePerPage>
          <TablePerPageLabel>Items per page</TablePerPageLabel>
          <TablePerPageSelector
            id="itemsPerPageSelector"
            autoCompleteItems={itemsPerPageOptions || []}
            inputProps={{
              value: pageLimit,
            }}
            onItemSelected={(item) => {
              const option = item as BaseItemType;
              setPageLimit(+option.id);
              return isFunction(onSetItemsPerPage)
                ? onSetItemsPerPage(option)
                : () => null;
            }}
          />
        </TablePerPage>
      );
    }

    return "";
  }, [showGoToLastPage, onGoToLastPageClicked, goToLastPageContent]);

  return (
    <TFoot>
      <tr>
        <td colSpan={2}>{itemsPerPageSelector}</td>
        <td colSpan={100}>
          {children}
          {itemsCount}
          {goToFirstPage}
          {goToPreviousPage}
          {goToNextPage}
          {goToLastPage}
        </td>
      </tr>
    </TFoot>
  );
};
