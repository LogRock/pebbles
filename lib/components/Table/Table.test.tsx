import React from "react";
import { render } from "../../utils/test-utils";
import Table, {
  OrderingEnum,
  TableBody,
  TableDatum,
  TableHead,
  TableHeader,
  TableRow,
} from "./";

describe("Table", () => {
  it("renders with some content", () => {
    expect(
      render(
        <Table>
          <TableHead>
            <TableRow addEmptyTDAtEnd>
              <TableHeader>Header A</TableHeader>
              <TableHeader showOrdering>Header A</TableHeader>
              <TableHeader showOrdering currentOrder={OrderingEnum.ASC}>
                Header B
              </TableHeader>
              <TableHeader showOrdering currentOrder={OrderingEnum.DESC}>
                Header C
              </TableHeader>
            </TableRow>
          </TableHead>
          <TableBody>
            <TableRow addEmptyTDAtEnd>
              <TableDatum>Content A1</TableDatum>
              <TableDatum>Content B1</TableDatum>
              <TableDatum>Content C1</TableDatum>
            </TableRow>

            <TableRow isNew>
              <TableDatum>Content A1</TableDatum>
              <TableDatum>Content B1</TableDatum>
              <TableDatum>Content C1</TableDatum>
            </TableRow>
          </TableBody>
        </Table>
      ).container.firstChild
    ).toMatchSnapshot();
  });

  it("renders selected rows", () => {
    expect(
      render(
        <Table>
          <TableHead>
            <TableRow>
              <TableHeader>Header A</TableHeader>
              <TableHeader>Header B</TableHeader>
              <TableHeader>Header C</TableHeader>
            </TableRow>
          </TableHead>
          <TableBody>
            <TableRow>
              <TableDatum>Content A1</TableDatum>
              <TableDatum>Content B1</TableDatum>
              <TableDatum>Content C1</TableDatum>
            </TableRow>

            <TableRow isSelected>
              <TableDatum>Content A2</TableDatum>
              <TableDatum>Content B2</TableDatum>
              <TableDatum>Content C2</TableDatum>
            </TableRow>

            <TableRow>
              <TableDatum>Content A3</TableDatum>
              <TableDatum>Content B3</TableDatum>
              <TableDatum>Content C3</TableDatum>
            </TableRow>
          </TableBody>
        </Table>
      ).container.firstChild
    ).toMatchSnapshot();
  });
});
