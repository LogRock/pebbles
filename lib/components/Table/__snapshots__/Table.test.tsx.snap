// Vitest Snapshot v1

exports[`Table > renders selected rows 1`] = `
<table
  class="sc-bqiRlB iddtuG"
>
  <thead
    class="sc-ksdxgE cKrAHV"
  >
    <tr
      class="sc-ieecCq kjoXAI"
    >
      <th
        class="sc-fFeiMQ cGlYYl"
      >
        Header A
      </th>
      <th
        class="sc-fFeiMQ cGlYYl"
      >
        Header B
      </th>
      <th
        class="sc-fFeiMQ cGlYYl"
      >
        Header C
      </th>
    </tr>
  </thead>
  <tbody
    class="sc-hBUSln QoxkZ"
  >
    <tr
      class="sc-ieecCq kjoXAI"
    >
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content A1
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content B1
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content C1
      </td>
    </tr>
    <tr
      class="sc-ieecCq hDJWbS"
    >
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content A2
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content B2
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content C2
      </td>
    </tr>
    <tr
      class="sc-ieecCq kjoXAI"
    >
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content A3
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content B3
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content C3
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table > renders with some content 1`] = `
<table
  class="sc-bqiRlB iddtuG"
>
  <thead
    class="sc-ksdxgE cKrAHV"
  >
    <tr
      class="sc-ieecCq kjoXAI"
    >
      <th
        class="sc-fFeiMQ cGlYYl"
      >
        Header A
      </th>
      <th
        class="sc-fFeiMQ cGlYYl"
      >
        Header A
        <svg
          fill="none"
          height="16"
          style="vertical-align: middle;"
          viewBox="0 0 11 16"
          width="11"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.375 5L5.375 1L9.375 5M9.375"
            stroke="#9CA3AF"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.67"
          />
          <path
            d="M1.375 5M9.375 11L5.375 15L1.375 11"
            stroke="#9CA3AF"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.67"
          />
        </svg>
      </th>
      <th
        class="sc-fFeiMQ cGlYYl"
      >
        Header B
        <svg
          fill="none"
          height="7"
          viewBox="0 0 12 7"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 6L6 1L1 6"
            stroke="#6B7280"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.67"
          />
        </svg>
      </th>
      <th
        class="sc-fFeiMQ cGlYYl"
      >
        Header C
        <svg
          fill="none"
          height="7"
          viewBox="0 0 12 7"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 1L6 6L1 1"
            stroke="#6B7280"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.67"
          />
        </svg>
      </th>
      <td
        class="sc-bkkeKt iKZsIB"
      />
    </tr>
  </thead>
  <tbody
    class="sc-hBUSln QoxkZ"
  >
    <tr
      class="sc-ieecCq kjoXAI"
    >
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content A1
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content B1
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content C1
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      />
    </tr>
    <tr
      class="sc-ieecCq ccTDHB"
    >
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content A1
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content B1
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        Content C1
      </td>
      <td
        class="sc-bkkeKt iKZsIB"
      >
        <span
          class="sc-dJjYzT cSXeVl"
        >
          New
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
